provider/safedns/safedns.go

```diff
@@ -0,0 +1,239 @@
+		return nil, err
+	}
+
+	// Check each found zone to see whether they match the domain filter provided. If they do, append it to the array of
+	// zones defined above. If not, continue to the next item in the loop.
+	for _, zone := range allZones {
+		if p.domainFilter.Match(zone.Name) {
+			zones = append(zones, zone)
+		} else {
+			continue
+		}
+	}
+	return zones, nil
+}
+
+func (p *SafeDNSProvider) ZoneRecords(ctx context.Context) ([]ZoneRecord, error) {
+	zones, err := p.Zones(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var zoneRecords []ZoneRecord
+	for _, zone := range zones {
+		// For each zone in the zonelist, get all records of an ExternalDNS supported type.
+		records, err := p.Client.GetZoneRecords(zone.Name, p.APIRequestParams)
+		if err != nil {
+			return nil, err
+		}
+		for _, r := range records {
+			zoneRecord := ZoneRecord{
+				ID:      r.ID,
+				Name:    r.Name,
+				Type:    r.Type,
+				TTL:     r.TTL,
+				Zone:    zone.Name,
+				Content: r.Content,
+			}
+			zoneRecords = append(zoneRecords, zoneRecord)
+		}
+	}
+	return zoneRecords, nil
+}
+
+// Records returns a list of Endpoint resources created from all records in supported zones.
+func (p *SafeDNSProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	var endpoints []*endpoint.Endpoint
+	zoneRecords, err := p.ZoneRecords(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for _, r := range zoneRecords {
+		if provider.SupportedRecordType(string(r.Type)) {
+			endpoints = append(endpoints, endpoint.NewEndpointWithTTL(r.Name, string(r.Type), endpoint.TTL(r.TTL), r.Content))
+		}
+	}
+	return endpoints, nil
+}
+
+// ApplyChanges applies a given set of changes in a given zone.
+func (p *SafeDNSProvider) ApplyChanges(ctx context.Context, changes *plan.Changes) error {
+	// Identify the zone name for each record
+	zoneNameIDMapper := provider.ZoneIDName{}
+
+	zones, err := p.Zones(ctx)
+	if err != nil {
+		return err
+	}
+	for _, zone := range zones {
+		zoneNameIDMapper.Add(zone.Name, zone.Name)
+	}
+
+	zoneRecords, err := p.ZoneRecords(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, endpoint := range changes.Create {
+		_, ZoneName := zoneNameIDMapper.FindZone(endpoint.DNSName)
+		for _, target := range endpoint.Targets {
+			request := safedns.CreateRecordRequest{
+				Name:    endpoint.DNSName,
+				Type:    endpoint.RecordType,
+				Content: target,
+			}
+			log.WithFields(log.Fields{
+				"zoneID":     ZoneName,
+				"dnsName":    endpoint.DNSName,
+				"recordType": endpoint.RecordType,
+				"Value":      target,
+			}).Info("Creating record")
+			_, err := p.Client.CreateZoneRecord(ZoneName, request)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	for _, endpoint := range changes.UpdateNew {
+		// TODO: Find a more effient way of doing this.
+		// Currently iterates over each zoneRecord in ZoneRecords for each Endpoint in UpdateNew; the same will go for
+		// Delete. As it's double-iteration, that's O(n^2), which isn't great.
```

Comment:
Should we be concerned about this? Is it something we are okay to merge like it is and then reconsider in the future?

Reply:
Personally, I haven't noticed any issues running this code in production for an account with many hundred of zones, each with many hundreds of records. On that basis alone, I'd say that it's okay to merge now and reconsider in future.

Reply:
Ok sounds good. I would prefer if we leave the TODO then if we do not need any change for now, so that we can search TODOs for real things that are left to do in the codebase.